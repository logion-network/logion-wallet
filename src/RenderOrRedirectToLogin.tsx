import { DateTime } from 'luxon';
import { Navigate, useLocation } from 'react-router-dom';

import { LOGIN_PATH, LocationState } from './Login';
import { useLogionChain } from './logion-chain';

export interface Props {
    render: () => JSX.Element;
}

export default function RenderOrRedirectToLogin(props: Props) {
    const { client } = useLogionChain();
    const location = useLocation();

    if(!client) {
        return null;
    }

    if(!client.isTokenValid(DateTime.now())) {
        const state: LocationState = {
            referrer: location.pathname
        };
        return <Navigate to={ LOGIN_PATH } state={ state } />;
    } else {
        return props.render();
    }
}
