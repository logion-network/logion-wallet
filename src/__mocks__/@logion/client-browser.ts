export class BrowserFile {

    constructor(blob: Blob) {
        this.blob = blob;
    }

    private blob: Blob;

    getBlob() {
        return this.blob;
    }
}

export class BrowserAxiosFileUploader {

    async upload() {

    }
}
