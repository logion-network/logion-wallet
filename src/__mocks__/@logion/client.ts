import { CheckResultType } from "@logion/client/dist/CollectionItem.js";
import { DefaultSignAndSendStrategy } from "@logion/client/dist/Signer.js";

import {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    LocRequestState,
    LocsState,
    HashOrContent,
} from '../LogionClientMock';

export {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    CheckResultType,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    DefaultSignAndSendStrategy,
    LocRequestState,
    LocsState,
    HashOrContent,
}

export { MimeType } from "@logion/client/dist/Mime.js";
