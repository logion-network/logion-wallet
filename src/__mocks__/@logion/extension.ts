import {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
} from '../PolkadotExtensionDappMock';
export {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
};
