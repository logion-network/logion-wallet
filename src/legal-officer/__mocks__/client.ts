import {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
} from "./ClientMock";

export {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
}
