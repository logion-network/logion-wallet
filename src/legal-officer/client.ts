import {
    EditableRequest,
    Signer,
    SignCallback,
    LocRequestState,
    OpenLoc,
    ClosedLoc,
    ClosedCollectionLoc,
    VerifiedThirdParty,
    VoidedLoc,
    VoidedCollectionLoc,
    LogionClient,
    LegalOfficer,
} from "@logion/client";
import {
    UUID,
    addLink as nodeApiAddLink,
    addFile as nodeApiAddFile,
    addMetadata as nodeApiAddMetadata,
    MetadataItem,
    closeLoc as nodeApiCloseLoc,
    voidLoc as nodeApiVoidLoc,
    VoidInfo,
    asString,
} from "@logion/node-api";
import { AnyJson } from "@polkadot/types-codec/types/helpers.js";
import { fetchAllLocsParams } from "../loc/LegalOfficerLocContext";

export async function addLink(params: {
    locState: EditableRequest,
    target: UUID,
    nature: string,
}): Promise<EditableRequest> {
    const { locState, target, nature } = params;
    const axios = buildAxios(locState);
    await axios.post(`/api/loc-request/${ locState.data().id.toString() }/links`, { target: target.toString(), nature })
    return await locState.refresh() as EditableRequest;
}

export async function deleteLink(params: {
    locState: EditableRequest,
    target: UUID,
}): Promise<EditableRequest> {
    const { locState, target } = params;
    const axios = buildAxios(locState);
    await axios.delete(`/api/loc-request/${ locState.data().id.toString() }/links/${target}`);
    return await locState.refresh() as EditableRequest;
}

function buildAxios(locState: LocRequestState) {
    const client = locState.locsState().client;
    return client.buildAxios(getLegalOfficer(client, locState.data().ownerAddress));
}

function getLegalOfficer(client: LogionClient, address: string): LegalOfficer {
    return client.legalOfficers.find(legalOfficer => address === legalOfficer.address)!;
}

export async function publishLink(params: {
    locState: EditableRequest,
    target: UUID,
    nature: string,
    signer: Signer,
    callback: SignCallback,
}): Promise<EditableRequest> {
    const { locState, target, nature, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locId = locState.data().id;
    const submittable = nodeApiAddLink({
        locId,
        api,
        target,
        nature,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${ locId.toString() }/links/${ target.toString() }/confirm`);

    return await locState.refresh() as EditableRequest;
}

export async function publishFile(params: {
    locState: EditableRequest,
    hash: string,
    nature: string,
    submitter: string,
    signer: Signer,
    callback: SignCallback,
}): Promise<EditableRequest> {
    const { locState, hash, nature, submitter, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locId = locState.data().id;
    const submittable = nodeApiAddFile({
        locId,
        api,
        hash,
        nature,
        submitter,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${ locId.toString() }/files/${ hash }/confirm`);

    return await locState.refresh() as EditableRequest;
}

export async function publishMetadata(params: {
    locState: EditableRequest,
    item: MetadataItem,
    signer: Signer,
    callback: SignCallback,
}): Promise<EditableRequest> {
    const { locState, item, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locId = locState.data().id;
    const submittable = nodeApiAddMetadata({
        locId,
        api,
        item,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${ locId.toString() }/metadata/${ encodeURIComponent(item.name) }/confirm`);

    return await locState.refresh() as EditableRequest;
}

export async function closeLoc(params: {
    locState: OpenLoc,
    signer: Signer,
    callback: SignCallback,
}): Promise<ClosedLoc | ClosedCollectionLoc> {
    const { locState, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locData = locState.data();
    const locId = locData.id;
    const seal = locData.seal;
    const submittable = nodeApiCloseLoc({
        locId,
        api,
        seal,
    });
    const signerId = locState.data().ownerAddress;
    await signer.signAndSend({
        signerId,
        submittable,
        callback
    });

    let refreshedLocState;
    if(locState.discarded) {
        refreshedLocState = (await client.locsState(fetchAllLocsParams(getLegalOfficer(client, signerId)))).findById(locId);
    } else {
        refreshedLocState = locState;
    }

    const axios = buildAxios(refreshedLocState);
    await axios.post(`/api/loc-request/${ locId.toString() }/close`);

    return await refreshedLocState.refresh() as ClosedLoc | ClosedCollectionLoc;
}

export interface FullVoidInfo extends VoidInfo {
    reason: string;
}

export async function voidLoc(params: {
    locState: LocRequestState,
    voidInfo: FullVoidInfo,
    signer: Signer,
    callback: SignCallback,
}): Promise<VoidedLoc | VoidedCollectionLoc> {
    const { locState, voidInfo, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locData = locState.data();
    const locId = locData.id;
    const submittable = nodeApiVoidLoc({
        locId,
        api,
        voidInfo,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    const reason = voidInfo.reason;
    await axios.post(`/api/loc-request/${ locId.toString() }/void`, {
        reason
    });

    return await locState.refresh() as VoidedLoc | VoidedCollectionLoc;
}

export async function setVerifiedThirdParty(params: {
        locState: ClosedLoc,
        isVerifiedThirdParty: boolean;
    }
): Promise<ClosedLoc> {
    const { locState, isVerifiedThirdParty } = params;
    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${locState.data().id.toString()}/verified-third-party`, {
        isVerifiedThirdParty
    });
    return await locState.refresh() as ClosedLoc;
}

export async function getVerifiedThirdPartySelections(params: { locState: OpenLoc } ): Promise<VerifiedThirdParty[]> {
    const { locState } = params
    const axios = buildAxios(locState);
    const response = await axios.get("/api/verified-third-parties");
    const allVerifiedThirdParties: VerifiedThirdParty[] = response.data.verifiedThirdParties;

    const selectedParties = locState.data().selectedParties;

    return allVerifiedThirdParties
        .filter(vtp => vtp.address !== locState.data().requesterAddress)
        .map(vtp => selectedParties.find(selectedParty => selectedParty.identityLocId === vtp.identityLocId && selectedParty.selected) ?
            { ...vtp, selected: true } :
            { ...vtp, selected: false }
        )
        .sort((vtp1, vtp2) => vtp1.lastName.localeCompare(vtp2.lastName));
}

export interface SelectPartiesParams {
    locState: OpenLoc;
    partyId: string;
}

export async function selectParties(params: SelectPartiesParams): Promise<void> {
    const { locState, partyId } = params;
    const axios = buildAxios(locState);
    await axios.post(`/api/loc-request/${locState.locId.toString()}/selected-parties`, {
        identityLocId: partyId
    })
}

export async function unselectParties(params: SelectPartiesParams): Promise<void> {
    const { locState, partyId } = params;
    const axios = buildAxios(locState);
    await axios.delete(`/api/loc-request/${ locState.locId.toString() }/selected-parties/${ partyId }`)
}

export async function requestVote(params: {
    locState: ClosedLoc | ClosedCollectionLoc,
    signer: Signer,
    callback: SignCallback,
}): Promise<string> {
    const { locState, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locData = locState.data();
    const locId = locData.id;
    const submittable = api.tx.vote.createVoteForAllLegalOfficers(locId.toDecimalString());
    const result = await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });
    const voteCreated = result.events.find(event => event.name === "VoteCreated" && event.section === "vote");
    if(!voteCreated) {
        throw new Error("Unable to retrieve vote ID");
    }
    const voteCreatedData = voteCreated.data as AnyJson[];
    return asString(voteCreatedData[0]);
}
