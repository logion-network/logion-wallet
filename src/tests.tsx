import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };

export async function clickByName(name: string | RegExp | ((accessibleName: string, element: Element) => boolean)) {
    let button: HTMLElement;
    await waitFor(() => button = screen.getByRole("button", { name }));
    await userEvent.click(button!);
}

export async function typeByLabel(label: string, value: string) {
    let input: HTMLElement;
    await waitFor(() => input = screen.getByLabelText(label));
    await userEvent.type(input!, value);
}
