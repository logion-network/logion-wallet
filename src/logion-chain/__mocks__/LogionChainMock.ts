import { DateTime } from "luxon";
import { LegalOfficer } from '@logion/client';
import { DEFAULT_LEGAL_OFFICER, legalOfficers } from 'src/common/TestData';
import { TEST_WALLET_USER } from 'src/wallet-user/TestData';
import { AxiosInstance } from 'axios';
import { Mock } from 'moq.ts';
import Accounts, { Account } from 'src/common/types/Accounts';
import { LogionClient } from '@logion/client/dist/LogionClient.js';
import { LogionClient as LogionClientMock } from '../../__mocks__/LogionClientMock';
import { api } from "src/__mocks__/LogionMock";
import { Call } from "../LogionChainContext";

export const LogionChainContextProvider = (props: any) => null;

export let clientMock: LogionClient | LogionClientMock | null = new LogionClientMock();

export let authenticate = jest.fn();

export let saveOfficer = jest.fn();

function getOfficer(address: string | undefined): LegalOfficer | undefined {
    return legalOfficers.find(legalOfficer => legalOfficer.address === address);
}

export let selectAddress = jest.fn();

export const DEFAULT_USER_ACCOUNT: Account = {
    name: "name",
    accountId: TEST_WALLET_USER,
    isLegalOfficer: false,
    token: {
        value: "token",
        expirationDateTime: DateTime.now(),
    },
}

export const DEFAULT_LEGAL_OFFICER_ACCOUNT: Account = {
    name: "name",
    accountId: DEFAULT_LEGAL_OFFICER,
    isLegalOfficer: true,
    token: {
        value: "token",
        expirationDateTime: DateTime.now(),
    },
}

export let accounts: Accounts | null = {
    current: DEFAULT_USER_ACCOUNT,
    all: [ DEFAULT_USER_ACCOUNT ]
}

export function setCurrentAddress(value: Account | undefined) {
    if(value === undefined) {
        accounts = {
            current: undefined,
            all: [ DEFAULT_USER_ACCOUNT ]
        };
    } else {
        accounts = {
            current: value,
            all: [ value ]
        };
    }
    if(context) {
        context.accounts = accounts;
    }
}

export function setAddresses(value: Accounts | null) {
    accounts = value;
}

export let axiosMock = new Mock<AxiosInstance>();

export function resetAxiosMock() {
    axiosMock = new Mock<AxiosInstance>();
}

let context: any = undefined;

export function setContextMock(value: any) {
    context = value;
}

let axiosFactory = () => axiosMock.object();

let signAndSend = jest.fn().mockResolvedValue(undefined);

let signer = {
    signAndSend,
};

export const NO_SUBMISSION: unknown = {
    canSubmit: () => true,
    isSuccessful: () => false,
    isError: () => false,
    error: null,
    result: null,
};

let extrinsicSubmissionState: unknown = NO_SUBMISSION;

export const SUCCESSFUL_SUBMISSION: unknown = {
    canSubmit: () => true,
    isSuccessful: () => true,
    isError: () => false,
    error: null,
    result: {
        status: {
            isFinalized: true,
        }
    },
};

export const FAILED_SUBMISSION: unknown = {
    canSubmit: () => false,
    isSuccessful: () => false,
    isError: () => true,
    error: "error",
    result: {
        status: {
            isFinalized: false,
        }
    },
};

export const PENDING_SUBMISSION: unknown = {
    canSubmit: () => false,
    isSuccessful: () => false,
    isError: () => false,
    error: null,
    result: {
        status: {
            isFinalized: false,
        }
    },
};

export function setExtrinsicSubmissionState(state: unknown) {
    extrinsicSubmissionState = state;
}

export function useLogionChain() {
    if(context) {
        return context;
    } else {
        return {
            injectedAccounts: null,
            api: api.object(),
            connectedNodeMetadata: {

            },
            selectAddress,
            accounts,
            axiosFactory,
            authenticate,
            getOfficer,
            saveOfficer,
            client: clientMock,
            isCurrentAuthenticated: () => true,
            authenticateAddress,
            signer,
            backendConfig: {
                [DEFAULT_LEGAL_OFFICER.address]: {
                    iDenfy: false,
                }
            },
            extrinsicSubmissionState,
            resetSubmissionState: () => {},
            submitSignAndSubmit: () => {},
            submitCall: (call: Call) => call(() => {}),
        };
    }
}

function authenticateAddress() {
    return clientMock;
}

export function setClientMock(mock: LogionClient | null) {
    clientMock = mock;
}
