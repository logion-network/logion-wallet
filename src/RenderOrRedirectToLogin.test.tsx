jest.mock('react-router-dom');
jest.mock('./logion-chain');

import { shallowRender, act } from './tests';

import RenderOrRedirectToLogin from './RenderOrRedirectToLogin';
import { setClientMock } from './logion-chain/__mocks__/LogionChainMock';
import { LogionClient } from '@logion/client';

describe("RenderOrRedirectToLogin", () => {

    it('redirects to login with referrer if not logged in', () => {
        setClientMock(buildClientMock(false));
        const render = jest.fn();
        const tree = shallowRender( <RenderOrRedirectToLogin render={ render } /> );
        expect(tree).toMatchSnapshot();
    });

    it('renders if logged in', () => {
        setClientMock(buildClientMock(true));
        const render = jest.fn();
        act(() => { shallowRender( <RenderOrRedirectToLogin render={ render } /> )});
        expect(render).toBeCalled();
    });

    it('empty without client', () => {
        setClientMock(null);
        const render = jest.fn();
        const tree = shallowRender( <RenderOrRedirectToLogin render={ render } /> );
        expect(tree).toMatchSnapshot();
    });
});

function buildClientMock(authenticated: boolean): LogionClient {
    return {
        isTokenValid: () => authenticated,
    } as unknown as LogionClient;
}
