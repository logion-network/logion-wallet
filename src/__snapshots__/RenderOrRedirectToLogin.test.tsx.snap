// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderOrRedirectToLogin empty without client 1`] = `null`;

exports[`RenderOrRedirectToLogin redirects to login with referrer if not logged in 1`] = `
<Navigate
  state={
    Object {
      "referrer": "/path",
    }
  }
  to="/login"
/>
`;
